File: src/lib/padOverrides.js
export function setPadMaskOverride(pad, expansion_mm) {
  pad.mask_override = { expansion_mm }
}

export function setPadPasteOverride(pad, scaleOrOffset) {
  if (typeof scaleOrOffset === 'number') {
    pad.paste_override = { scale: scaleOrOffset }
  } else if (scaleOrOffset && typeof scaleOrOffset === 'object') {
    pad.paste_override = { ...scaleOrOffset }
  }
}

export function getEffectivePadMask(pad, board_defaults) {
  const expansion = pad.mask_override?.expansion_mm ??
    (board_defaults == null ? 0.05 : board_defaults.mask_expansion_mm !== undefined ? board_defaults.mask_expansion_mm : 0)
  const w = pad.width ?? pad.pad_diameter ?? 0
  const h = pad.height ?? pad.pad_diameter ?? 0
  const x = pad.x ?? 0
  const y = pad.y ?? 0
  return [
    [x - w / 2 - expansion, y - h / 2 - expansion],
    [x + w / 2 + expansion, y - h / 2 - expansion],
    [x + w / 2 + expansion, y + h / 2 + expansion],
    [x - w / 2 - expansion, y + h / 2 + expansion],
  ]
}

export function getEffectivePadPaste(pad, board_defaults) {
  const override = pad.paste_override
  const defaultScale = board_defaults?.paste_scale ?? 1.0

  if (override?.polygon) {
    return override.polygon
  }

  const scale = override?.scale ?? defaultScale
  const offset = override?.offset_mm ?? 0
  const w = pad.width ?? pad.pad_diameter ?? 0
  const h = pad.height ?? pad.pad_diameter ?? 0
  const x = pad.x ?? 0
  const y = pad.y ?? 0

  const sw = w * scale
  const sh = h * scale

  return [
    [x - sw / 2 + offset, y - sh / 2 + offset],
    [x + sw / 2 + offset, y - sh / 2 + offset],
    [x + sw / 2 + offset, y + sh / 2 + offset],
    [x - sw / 2 + offset, y + sh / 2 + offset],
  ]
}

export function validatePadOverrides(pad) {
  const errors = []
  if (!pad) {
    errors.push('pad is required')
    return errors
  }
  if (pad.mask_override !== undefined) {
    if (typeof pad.mask_override.expansion_mm !== 'number' || pad.mask_override.expansion_mm < 0) {
      errors.push('mask_override.expansion_mm must be a non-negative number')
    }
  }
  if (pad.paste_override !== undefined) {
    const po = pad.paste_override
    if (po.scale !== undefined && (typeof po.scale !== 'number' || po.scale < 0)) {
      errors.push('paste_override.scale must be a non-negative number')
    }
    if (po.offset_mm !== undefined && typeof po.offset_mm !== 'number') {
      errors.push('paste_override.offset_mm must be a number')
    }
    if (po.polygon !== undefined) {
      if (!Array.isArray(po.polygon) || po.polygon.length < 3) {
        errors.push('paste_override.polygon must be an array of at least 3 points')
      }
    }
  }
  return errors
}

File: packages/kerf-electronics/src/kerf_electronics/tools/pad_overrides.py
import json
from typing import Any

from kerf_electronics._compat import ToolSpec, err_payload, ok_payload, register


def _get_pad(circuit_json: dict, pad_id: str) -> dict | None:
    if isinstance(circuit_json, list):
        for el in circuit_json:
            if isinstance(el, dict) and el.get('pcb_smtpad_id') == pad_id:
                return el
        return None
    if isinstance(circuit_json, dict):
        if circuit_json.get('pcb_smtpad_id') == pad_id:
            return circuit_json
        pads = circuit_json.get('pcb_smtpad') or circuit_json.get('pcb_smtpad')
        if isinstance(pads, list):
            for pad in pads:
                if isinstance(pad, dict) and pad.get('pcb_smtpad_id') == pad_id:
                    return pad
    return None


set_pad_mask_override_spec = ToolSpec(
    name="set_pad_mask_override",
    description=(
        "Set or update the solder-mask aperture expansion for a specific SMT pad. "
        "Positive expansion_mm opens the mask window (removes mask), "
        "e.g. for a QFN exposed thermal pad where no solder mask should cover the pad. "
        "Pass expansion_mm=0 to suppress mask entirely if the pad's own mask_override is used."
    ),
    input_schema={
        "type": "object",
        "properties": {
            "circuit_json": {"type": "object"},
            "pad_id": {"type": "string"},
            "expansion_mm": {"type": "number"},
        },
        "required": ["circuit_json", "pad_id", "expansion_mm"],
    },
)


@register(set_pad_mask_override_spec, write=True)
async def set_pad_mask_override(ctx: Any, args: bytes) -> str:
    try:
        kwargs = json.loads(args)
    except Exception as e:
        return err_payload(str(e), "PARSE_ERROR")

    circuit_json = kwargs.get("circuit_json")
    pad_id = kwargs.get("pad_id")
    expansion_mm = kwargs.get("expansion_mm")

    if not isinstance(expansion_mm, (int, float)) or expansion_mm < 0:
        return err_payload("expansion_mm must be a non-negative number", "VALIDATION_ERROR")

    pad = _get_pad(circuit_json, pad_id)
    if pad is None:
        return err_payload(f"pad not found: {pad_id}", "NOT_FOUND")

    pad["mask_override"] = {"expansion_mm": float(expansion_mm)}
    return ok_payload({"circuit_json": circuit_json, "pad_id": pad_id, "expansion_mm": expansion_mm})


set_pad_paste_override_spec = ToolSpec(
    name="set_pad_paste_override",
    description=(
        "Set or update the solder-paste stencil aperture for a specific SMT pad. "
        "Use scale (< 1.0) to reduce paste coverage for fine-pitch parts. "
        "Use offset_mm to shift the aperture. "
        "Use polygon to define a fully custom aperture shape."
    ),
    input_schema={
        "type": "object",
        "properties": {
            "circuit_json": {"type": "object"},
            "pad_id": {"type": "string"},
            "scale": {"type": "number"},
            "offset_mm": {"type": "number"},
            "polygon": {
                "type": "array",
                "items": {"type": "array", "items": {"type": "number"}, "minItems": 2, "maxItems": 2},
            },
        },
        "required": ["circuit_json", "pad_id"],
    },
)


@register(set_pad_paste_override_spec, write=True)
async def set_pad_paste_override(ctx: Any, args: bytes) -> str:
    try:
        kwargs = json.loads(args)
    except Exception as e:
        return err_payload(str(e), "PARSE_ERROR")

    circuit_json = kwargs.get("circuit_json")
    pad_id = kwargs.get("pad_id")
    scale = kwargs.get("scale")
    offset_mm = kwargs.get("offset_mm")
    polygon = kwargs.get("polygon")

    if scale is not None and (not isinstance(scale, (int, float)) or scale < 0):
        return err_payload("scale must be a non-negative number", "VALIDATION_ERROR")
    if offset_mm is not None and not isinstance(offset_mm, (int, float)):
        return err_payload("offset_mm must be a number", "VALIDATION_ERROR")
    if polygon is not None:
        if not isinstance(polygon, list) or len(polygon) < 3:
            return err_payload("polygon must be an array of at least 3 [x, y] points", "VALIDATION_ERROR")

    pad = _get_pad(circuit_json, pad_id)
    if pad is None:
        return err_payload(f"pad not found: {pad_id}", "NOT_FOUND")

    override = {}
    if scale is not None:
        override["scale"] = float(scale)
    if offset_mm is not None:
        override["offset_mm"] = float(offset_mm)
    if polygon is not None:
        override["polygon"] = polygon

    pad["paste_override"] = override
    return ok_payload({"circuit_json": circuit_json, "pad_id": pad_id, "paste_override": override})


clear_pad_overrides_spec = ToolSpec(
    name="clear_pad_overrides",
    description="Remove mask_override and paste_override from a specific pad, reverting to board defaults.",
    input_schema={
        "type": "object",
        "properties": {
            "circuit_json": {"type": "object"},
            "pad_id": {"type": "string"},
        },
        "required": ["circuit_json", "pad_id"],
    },
)


@register(clear_pad_overrides_spec, write=True)
async def clear_pad_overrides(ctx: Any, args: bytes) -> str:
    try:
        kwargs = json.loads(args)
    except Exception as e:
        return err_payload(str(e), "PARSE_ERROR")

    circuit_json = kwargs.get("circuit_json")
    pad_id = kwargs.get("pad_id")

    pad = _get_pad(circuit_json, pad_id)
    if pad is None:
        return err_payload(f"pad not found: {pad_id}", "NOT_FOUND")

    if "mask_override" in pad:
        del pad["mask_override"]
    if "paste_override" in pad:
        del pad["paste_override"]

    return ok_payload({"circuit_json": circuit_json, "pad_id": pad_id})